Delint dockerfiles

wainersm opened the pull request:
I'm sending this PR in preparation to enable a checker on `static-checks.sh` which aims to [run a linter on the dockerfiles](https://github.com/kata-containers/tests/issues/4149) of the project.

That new checker will be using the [hadolint tool](https://github.com/hadolint/hadolint) to inspect the dockerfiles.  So the changes here were made to silent the warnings/errors pointed out by hadolint 2.7.0.

Note that we have a bunch of dockerfile templates on the project which can't be parsed by the tool without some tricks.  See on https://github.com/kata-containers/tests/pull/4194 how I plan to handle templates, but even with the handling some files fail to get parsed and that explain why they will be ignored.

Below is how I run the hadolint on kata-containers repository. 
```
$ curl -sL https://github.com/hadolint/hadolint/releases/download/v2.7.0/hadolint-Linux-x86_64 -o ~/go/bin/hadolint && chmod +x ~/go/bin/hadolint
$ cat dockerfile_linter.sh 
#!/bin/bash

cmd="$HOME/go/bin/hadolint --ignore DL3003 --ignore DL3048 --ignore DL3041 --ignore DL3033 --ignore DL3008 --ignore DL3018 --ignore DL3037 --failure-threshold warning"

ignore_files=(
	"tools/osbuilder/rootfs-builder/debian/Dockerfile-aarch64.in"
	"tools/osbuilder/rootfs-builder/centos/Dockerfile.in"
	"tools/osbuilder/rootfs-builder/debian/Dockerfile.in"
	"tools/osbuilder/rootfs-builder/fedora/Dockerfile.in"
        "tools/osbuilder/rootfs-builder/ubuntu/Dockerfile.in"
	"tools/osbuilder/rootfs-builder/ubuntu/Dockerfile-aarch64.in"
	"tools/packaging/tests/Dockerfile/Dockerfile.in"
	"tools/packaging/tests/Dockerfile/FedoraDockerfile.in"
	)
for f in `git ls-files | grep "Dockerfile" | grep -v '/vendor/' | grep -v 'template$'`; do
	echo "== Linting $f =="
	if echo "${ignore_files[@]}" | grep -q $f ; then
		echo 'skip'
		continue
	elif [[ "$f" =~ Dockerfile.*.in ]]; then
		sed -e 's/@[A-Z_]*@//' "$f" | $cmd - || exit 1
	else
		${cmd} $f || exit 1
	fi
done
$ ./dockerfile_linter.sh
== Linting ci/openshift-ci/images/Dockerfile.buildroot ==
== Linting tools/osbuilder/dockerfiles/QAT/Dockerfile ==
== Linting tools/osbuilder/dracut/Dockerfile.in ==
== Linting tools/osbuilder/image-builder/Dockerfile ==
== Linting tools/osbuilder/rootfs-builder/alpine/Dockerfile.in ==
-:9 DL3019 info: Use the `--no-cache` switch to avoid the need to use `--update` and remove `/var/cache/apk/*` when done installing packages
== Linting tools/osbuilder/rootfs-builder/centos/Dockerfile.in ==
skip
== Linting tools/osbuilder/rootfs-builder/clearlinux/Dockerfile.in ==
== Linting tools/osbuilder/rootfs-builder/debian/Dockerfile-aarch64.in ==
skip
== Linting tools/osbuilder/rootfs-builder/debian/Dockerfile.in ==
skip
== Linting tools/osbuilder/rootfs-builder/fedora/Dockerfile.in ==
skip
== Linting tools/osbuilder/rootfs-builder/gentoo/Dockerfile.in ==
== Linting tools/osbuilder/rootfs-builder/suse/Dockerfile.in ==
== Linting tools/osbuilder/rootfs-builder/ubuntu/Dockerfile-aarch64.in ==
skip
== Linting tools/osbuilder/rootfs-builder/ubuntu/Dockerfile.in ==
skip
== Linting tools/packaging/kata-deploy/Dockerfile ==
tools/packaging/kata-deploy/Dockerfile:39 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
== Linting tools/packaging/kata-deploy/action/Dockerfile ==
tools/packaging/kata-deploy/action/Dockerfile:23 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
== Linting tools/packaging/kata-deploy/local-build/dockerbuild/Dockerfile ==
tools/packaging/kata-deploy/local-build/dockerbuild/Dockerfile:26 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
tools/packaging/kata-deploy/local-build/dockerbuild/Dockerfile:27 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
== Linting tools/packaging/kata-monitor/Dockerfile ==
== Linting tools/packaging/static-build/kernel/Dockerfile ==
== Linting tools/packaging/static-build/qemu/Dockerfile ==
tools/packaging/static-build/qemu/Dockerfile:15 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
== Linting tools/packaging/static-build/shim-v2/Dockerfile ==
== Linting tools/packaging/tests/Dockerfile/Dockerfile.in ==
skip
== Linting tools/packaging/tests/Dockerfile/FedoraDockerfile.in ==
skip
```

wainersm commented:
Updated the https://github.com/kata-containers/kata-containers/pull/3109/commits/899345a40e6d41d2a88996df659849977b779bbd to fix a problem pointed out by CI. There are dockerfiles now passing `--no-install-recommends` to `apt` and essential packages like `ca-certificates` aren't installed.

Bevisy commented on tools/packaging/static-build/qemu/Dockerfile:
```diff
@@ -46,40 +46,37 @@ RUN apt-get --no-install-recommends install -y \
+COPY static-build/qemu.blacklist /root/static-build/qemu.blacklist
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN git clone "${QEMU_REPO}" qemu && \
```
We can reduce the size of the pull code and increase the speed of the build by adding the parameter "-- depth=1" for all repos used here.

wainersm replied:
Good idea @Bevisy ! 

BTW, If I am not wrong QEMU provides a script to allow cloning/updating the sub-modules trees. Maybe we should use that instead of cloning manually.

wainersm replied:
Done in https://github.com/kata-containers/kata-containers/issues/3291

Bevisy commented on tools/osbuilder/rootfs-builder/debian/Dockerfile-aarch64.in:
```diff
@@ -8,7 +8,7 @@ ARG IMAGE_REGISTRY=docker.io
 
 # RUN commands
-RUN apt-get update && apt-get install -y \
+RUN apt-get update && apt-get install -y --no-install-recommends \
```
Non-advice. Just want to express that this has taught me a lot :)

wainersm replied:
Me either!

Bevisy commented on tools/packaging/tests/Dockerfile/FedoraDockerfile.in:
```diff
@@ -14,15 +14,14 @@ ENV GOPATH=/home/go
+    mkdir "${GOPATH}" && \
+    dnf config-manager --add-repo https://download.docker.com/linux/fedora/docker-ce.repo && \
+    dnf makecache && dnf -y install docker-ce && dnf clean all && \
+    go get -d "${TESTS_REPO}" && \
```
We can change "go get -d" to "git clone --depth=1" for reducing the size of the repo and increasing the building speed

wainersm commented:
/test

wainersm commented:
Hi @egernst , the static-checker is complaining about the missing copyright statement in `tools/packaging/kata-monitor/Dockerfile` (file added in commit 2f1cb7995ffe8089ea3c0127736d3f0bcd711fc9). What copyright statement can I place there?

wainersm commented on tools/osbuilder/image-builder/Dockerfile:
```diff
@@ -5,6 +5,7 @@
-RUN dnf install -y qemu-img parted gdisk e2fsprogs gcc xfsprogs findutils
+RUN ([ -n "$http_proxy" ] && \
+    sed -i '$ a proxy='$http_proxy /etc/dnf/dnf.conf ; true) && \
+    dnf install -y qemu-img parted gdisk e2fsprogs gcc xfsprogs findutils && \
```
done.

wainersm commented:
Updated this PR:
 * Rebased and resolved conflicts
 * Added the `ca-certificates` package in `tools/packaging/static-build/kernel/Dockerfile`. I hope it will make the kata deploy build CI to run fine
 * Addressed a comment from @Bevisy

wainersm commented:
Hi @fidencio ! Do you have some cycles to review this PR?

In particular I made changes to the kata-deploy  that I'd like to double-check with you. The `azure-cli` has now a pined version, but if `latest` known to no cause trouble then I can change that.

wainersm commented:
/test

Jakob-Naucke commented:
@wainersm PTAL at the static failures

wainersm commented:
> @wainersm PTAL at the static failures

Yep, it complains about the lack of copyrights on tools/packaging/kata-monitor/Dockerfile . That file was introduced by @egernst so I need him to tell me the proper statement.

wainersm commented:
/test-arm

wainersm commented:
/test-s390x

Jakob-Naucke commented:
/retest-s390x

wainersm commented:
Updated this PR with:
 * rebased to fix conflicts
 * added https://github.com/kata-containers/kata-containers/pull/3109/commits/d79268ac6519088b21011772a61337bcc9930699 to solve the static-checker fail

wainersm commented:
/test

jodh-intel commented on ci/openshift-ci/images/Dockerfile.buildroot:
```diff
@@ -6,4 +6,9 @@
+    yum -y install \
+    git \
+    sudo \
+    wget && \
```
Nit: We should really be using `curl`, but not strictly required by this PR (but you could add another commit).

jodh-intel reviewed (approved):
Thanks @wainersm.

lgtm

jodh-intel commented:
/retest-arm
/retest-vfio
